src/toil/jobStores/fileJobStore.py

```diff
@@ -121,8 +122,18 @@ def batch(self):
             self.update(jobGraph)
         self._batchedJobGraphs = None
 
-    def exists(self, jobStoreID):
-        return os.path.exists(self._getJobFileName(jobStoreID))
+    def exists(self, jobStoreID,maxTries=1):
+        sleepTime = 1
+        for iTry in range(1,maxTries+1):
+            jobFile = self._getJobFileName(jobStoreID)
+            if os.path.exists(jobFile):
+                return True
+            if iTry >= maxTries:
+                return False
+            logger.warning("Job file `{}` for job `{}` does not exist, retry #{} in {}s.".\
+                    format(jobFile,jobStoreID,iTry+1,sleepTime))
+            time.sleep(sleepTime)
```

Comment:
ah, I see this is on  NFS.  This could be an issue with NFS caching.  I can be a real pain.  I might make sense to make the JobStore NFS-aware.

Reply:
Totally agree that spin-waiting can be hiding bugs in the code. It did help with persistent tracebacks when running on the NFS, though. Can you tell me: is it possible in Toil design that the leader is checking for the existence of a job file that is supposed to be created on a different node (by a worker)? It is hard to figure it out by looking at the code. If the answer is Yes, and the shared FS is essentially used for synchronization, then this strongly supports the need for spin-waiting on the NFS. On our cluster, it usually takes about 15 sec for the files to appear in `ls` across nodes. IT played with various settings on the NetApp and nodes trying to make this delay shorter, and to no avail. Various other workflow systems have settings to control how long they should wait for the outputs after completion of a task.

Reply:
IIRC, FileJobStore assumes Posix semantics. If the underlying file-system is only eventually consistent–and it appears that NFS is–then the need arises for another strongly consistent mechanism to tell the job store that a file will, definitely, but eventually appear. This is how the AWS job store works, it uses the strongly consistent SimpleDB for tracking which files exist. Merely, spin-locking would not work as it would introduce the possibility of dead lock: Toil relies on both the absence and the presence of files to trigger a progression in the workflow, I think.

A hacky workaround would be, as @andreyto suggests, to always wait for a configurable time for a file to appear or to disappear (!) but this would slow down the workflow. Maybe the disappearance delay can be limited to only those cases in which workflow progress is gated on the absence of a file.

Reply:
The spin-wait seems a necessary evil to support NFS without a lot of investment.  NFS can be run in a coherent manner with at a significant and often unacceptable slowdown.

My concerns about this are
  - are there any other places where this kind of spin-wait is necessary?
  - do we need a config variable for the time?
  - the rationale is not documented in the code for those, like me, who have forgotten the pain of NFS.

Reply:
I like @diekhans points. A config variable and better documentation (rationale) should be added.

For the other point, my understanding is that the fileJobStore was built to test workflows locally. There could well be other consistency issues. Ideally, these would be handled in an abstraction layer since this problem applies to the other jobStores as @hannes-ucsc pointed out.

Does anyone object to patching it until we have time for a better solution.

Reply:
@hannes-ucsc , could you describe and/or point to places in the code where `Toil relies on <snip> the absence `?
As far as NFS consistency guarantees, they are described in the `man nfs` help page, as well as discussed in [this thread](https://serverfault.com/questions/638601/nfs-caching-issue).
In short, without config changes that kill the performance, the NFS has:
- no cache coherency between clients for open file writes/reads.  MPI applications with concurrent I/O to the same file are affected by this, and that leads to the development of parallel file systems like GPFS, Lustre, PVFS. Toil is not affected by this because the client/server pipe on the NFS always does full sync when the file is closed, and Toil does not do concurrent writes (I hope).
- no cache coherency for metadata operations (directory modifications). Toil is affected by this, but not all is bad. NFS (see the man page) provides strict time guarantees for how long it will take for the cache to be reconciled on the client. These can be tuned in `/etc/fstab`. Incidentally, the 30 sec max wait that I set in the spin-wait code is the default in the NFS man page. Most sites never change the default.

A number of workflow systems implement the spin-wait for this very pattern (Makeflow, Galaxy, Cromwell), and do just fine. Basically, the system learns that the job has finished when its disappears from `qstat` output, and the system waits for the output files to appear on the NFS. Consider also that the batch job scheduling system has its own delays in starting jobs, so the extra worst-case 30 sec (with default NFS setup) of wait will be incremental to the queue delays (probably double the delay).

It is possible of course to implement some other notification mechanism like running a 0MQ server process in the leader, but is it really worth the trouble?

I agree that the spin-wait time should be made a tunable parameter.

Reply:
> could you describe and/or point to places in the code where

https://github.com/BD2KGenomics/toil/blob/27edc4b64e992f29de64c863f00044771d494022/src/toil/leader.py#L834

is a place where the deletion of a file is used as a signal. Or

https://github.com/BD2KGenomics/toil/blob/27edc4b64e992f29de64c863f00044771d494022/src/toil/leader.py#L732

is an important condition where the absence of a file moves the workflow forward.

And this is just the result of a cursory scan of the code base. I think every job store file access would have to be scrutinized for the effects of eventual consistency.

> NFS (see the man page) provides strict time guarantees 

Whenever I hear the words "NFS", "strict" and "guarantee" in the same sentence, I smile. But maybe NFS has gotten better. I haven't used it in a while.

> A number of workflow systems implement the spin-wait for this very pattern (Makeflow, Galaxy, Cromwell), 

Do they also use the file system to manage the workflow state? As I mentioned before: if you know through an out-of-band channel that a file *should* (*should not*) exist, then its trivial to just wait until it appears (disappears). But Toil tracks the workflow itself in the job store by persisting the very job objects to it, not just the inputs and outputs of those jobs.

> running a 0MQ server process in the leader, but is it really worth the trouble

Toil workflows are resumable from nothing more but the information in the job store, so relying on ephemeral messages to communicate information about files out of band may be difficult.

I want to caution against fixing one particular workflow to deal with NFS's eventual consistency and concluding that that fixes every other workflow on every other installation of NFS.

Maybe the Toil team should experiment with an eventually consistent file system, or at least emulate one and run the job store unit tests on it. The difficulty lies in the fact that NFS's eventual consistency can only be observed with more than one node.

Reply:
By the way, I don't speak for the Toil team. I'm just a former contributor. The above is merely my own opinion.